reco_utils/recommender/lightgbm/lightgbm_utils.py

```diff
@@ -0,0 +1,224 @@
+    remote_path = azure_container_url + remote_resource_name
+    maybe_download(remote_path, remote_resource_name, data_path)
+    tar_ref = tarfile.open(os.path.join(data_path, remote_resource_name), "r")
+    tar_ref.extractall(data_path)
+    tar_ref.close()
+    os.remove(os.path.join(data_path, remote_resource_name))
+
+def cal_metric(labels, preds, metrics):
+    """Calculate metrics,such as auc, logloss
+    FIXME: refactor this with the reco metrics
+    """
+    res = {}
+    for metric in metrics:
+        if metric == "auc":
+            auc = roc_auc_score(np.asarray(labels), np.asarray(preds))
+            res["auc"] = round(auc, 4)
+        elif metric == "rmse":
+            rmse = mean_squared_error(np.asarray(labels), np.asarray(preds))
+            res["rmse"] = np.sqrt(round(rmse, 4))
+        elif metric == "logloss":
+            # avoid logloss nan
+            preds = [max(min(p, 1.0 - 10e-12), 10e-12) for p in preds]
+            logloss = log_loss(np.asarray(labels), np.asarray(preds))
+            res["logloss"] = round(logloss, 4)
+        elif metric == "acc":
+            pred = np.asarray(preds)
+            pred[pred >= 0.5] = 1
+            pred[pred < 0.5] = 0
+            acc = accuracy_score(np.asarray(labels), pred)
+            res["acc"] = round(acc, 4)
+        elif metric == "f1":
+            pred = np.asarray(preds)
+            pred[pred >= 0.5] = 1
+            pred[pred < 0.5] = 0
+            f1 = f1_score(np.asarray(labels), pred)
+            res["f1"] = round(f1, 4)
+        else:
+            raise ValueError("not define this metric {0}".format(metric))
+    return res
+
+def unpackbits(x,num_bits):
+    xshape = list(x.shape)
+    x = x.reshape([-1,1])
+    to_and = 2**np.arange(num_bits).reshape([1,num_bits])
+    return (x & to_and).astype(bool).astype(int).reshape(xshape + [num_bits])
+
+class NumEncoder(object):
+    def __init__(self, cate_cols, nume_cols, label_col, threshold=10, thresrate=0.99):
+        self.label_name = label_col
+        self.cate_cols = cate_cols
+        self.dtype_dict = {}
+        for item in cate_cols:
+            self.dtype_dict[item] = 'str'
+        for item in nume_cols:
+            self.dtype_dict[item] = 'float'
+        self.nume_cols = nume_cols
+        self.tgt_nume_cols = []
+        self.encoder = ce.ordinal.OrdinalEncoder(cols=cate_cols)
+        self.threshold = threshold
+        self.thresrate = thresrate
+        
+        self.save_cate_avgs = {}
+        self.save_value_filter = {}
+        self.save_num_embs = {}
+        self.Max_len = {}
+        self.samples = 0
+
+    def fit_transform(self, inPath):
+        logging.info('----------------------------------------------------------------------')
+        logging.info('Fitting and Transforming %s .'%inPath)
+        logging.info('----------------------------------------------------------------------')
+        df = pd.read_csv(inPath, dtype=self.dtype_dict)
```

Comment:
Do we want to accept a dataframe instead of a file path as input, which might make it easier for users that interface with a prepared data for model fitting and transforming?

Reply:
Fixed.